unpacker.go

```diff
@@ -0,0 +1,352 @@
+				break
+			}
+			if states[i].unpacked || states[i].empty {
+				continue
+			}
+			// unpack the layer
+			unpacked, err := rootfs.ApplyLayerWithOpts(ctx, states[i].layer, chain, sn, a,
+				u.config.SnapshotOpts, u.config.ApplyOpts)
+			if err != nil {
+				return nil, err
+			}
+
+			if unpacked {
+				// Set the uncompressed label after the uncompressed
+				// digest has been verified through apply.
+				cinfo := content.Info{
+					Digest: states[i].layer.Blob.Digest,
+					Labels: map[string]string{
+						"containerd.io/uncompressed": states[i].layer.Diff.Digest.String(),
+					},
+				}
+				if _, err := cs.Update(ctx, cinfo, "labels.containerd.io/uncompressed"); err != nil {
+					return nil, err
+				}
+			}
+
+			chain = append(chain, states[i].layer.Diff.Digest)
+			states[i].unpacked = true
+			log.G(ctx).WithFields(logrus.Fields{
+				"desc": states[i].layer.Blob,
+				"diff": states[i].layer.Diff,
+			}).Debug("layer unpacked")
+		}
+	}
+
+	for _, state := range states {
+		if !state.unpacked && !state.empty {
+			return nil, errors.Errorf("layer %q is not unpacked", state.layer.Blob.Digest)
+		}
+	}
+	log.G(ctx).WithFields(logrus.Fields{
+		"desc":  manifestDesc,
+		"chain": chain,
+	}).Debug("image unpacked")
+	return chain, nil
+}
+
+func (u *unpacker) handler(f images.Handler) images.HandlerFunc {
+	return func(ctx context.Context, desc ocispec.Descriptor) ([]ocispec.Descriptor, error) {
+		children, err := f.Handle(ctx, desc)
+		if err != nil {
+			return children, err
+		}
+
+		switch desc.MediaType {
+		case images.MediaTypeDockerSchema1Manifest:
+			if len(children) == 0 {
+				return children, errors.Errorf("unexpected children %v", children)
+			}
+			u.manifestCh <- manifestUpdate{
+				descriptor: desc,
+				children:   children,
+			}
+		case images.MediaTypeDockerSchema2Manifest, ocispec.MediaTypeImageManifest:
+			if len(children) == 0 {
+				return children, errors.Errorf("unexpected children %v", children)
+			}
+			var descs []ocispec.Descriptor
+			for _, child := range children {
+				if child.MediaType == images.MediaTypeDockerSchema2Config ||
+					child.MediaType == ocispec.MediaTypeImageConfig {
+					continue
+				}
+				descs = append(descs, child)
+			}
+			u.manifestCh <- manifestUpdate{
+				descriptor: desc,
+				children:   descs,
+			}
+		case images.MediaTypeDockerSchema2Config, ocispec.MediaTypeImageConfig:
+			p, err := content.ReadBlob(ctx, u.c.ContentStore(), desc)
+			if err != nil {
+				return children, err
+			}
+
+			var i ocispec.Image
+			if err := json.Unmarshal(p, &i); err != nil {
+				return children, err
+			}
+
+			var diffIDs []ocispec.Descriptor
+			for _, diffID := range i.RootFS.DiffIDs {
+				diffIDs = append(diffIDs, ocispec.Descriptor{
+					MediaType: ocispec.MediaTypeImageLayer,
+					Digest:    diffID,
+				})
+			}
+			u.manifestCh <- manifestUpdate{
+				descriptor: desc,
+				children:   diffIDs,
+			}
+		default:
+			u.layerCh <- desc
```

Comment:
I think we can do nothing for manifest list instead of sending invalid layer.

Reply:
We can skip manifest list.